### ISSUE TITLE
[BUG] Prediction tensors not moved to CPU between batches, causing VRAM accumulation

### ISSUE THREAD

**Issue by StrikerEureka34** [jdb78/pytorch-forecasting#2224]
**Describe the bug**

When running `model.predict()` on a CUDA device, `PredictCallback.on_predict_batch_end` collects each batch output into self.predictions without calling `.detach()` or `.cpu()`:

[Bug Lives Here:](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/callbacks/predict.py#L69)

```python
self.predictions.append(processed_output)
```

Every prediction tensor stays pinned in VRAM until `on_predict_epoch_end` runs at the very end of inference. The same issue affects `return_info` tensors at [here](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/callbacks/predict.py#L73-L79).

While reviewing I found that, PredictCallback was introduced in [#1984](https://github.com/sktime/pytorch-forecasting/pull/1984) as the v2 predict path and the v1 base model had an explicit guard at [here](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/models/base/_base_model.py#L310-L311) that moved outputs off-device after every batch:

```python
out = move_to_device(detach(out), "cpu")
x   = move_to_device(detach(x),   "cpu")
```

But this pattern was not carried over into the v2 callback.

&nbsp;

**To Reproduce**

Full repro script (any CUDA GPU): [Gist here](https://gist.github.com/StrikerEureka34/19f78514c1a3fbb39b7f353d67599f37)
Result I got while running it locally are:-

```python
# device: cuda
# VRAM before loop: 0.0 MB
# batches collected : 127
# tensor shape      : torch.Size([64, 24])
# tensor device     : cuda:0
# VRAM held by 127 prediction tensors: 9.44 MB
# Expected: ~0 MB (tensors should be on CPU). Got: 9.44 MB on GPU.
# With a real FM dataset (1000+ series, multivariate), this accumulates to OOM.
```

&ensp;

**Expected behavior**

So here `cb.predictions[0].device` should be cpu after each batch. But right now it is `cuda:0`, which means every batch tensor accumulates in VRAM for the entire inference run.

&ensp;

**Proposed Fix**

Add '.detach().cpu()' before each append at [here](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/callbacks/predict.py#L69) for predictions, and at [here](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/callbacks/predict.py#L73-L79) for return_info tensors (y, decoder_lengths) and a dict comprehension over x since it contains mixed tensor and non-tensor values.

&ensp;

**Additional context**

For FM-scale workloads(#1959), inference over thousands of series runs hundreds of batches per call. 
This accumulation is linear, for example 127 batches on this test already held 9.44 MB above model weight allocations. 

What is worse is that on a smaller GPU or with multivariate targets this quietly exhausts VRAM mid-inference with no message pointing to PredictCallback as the cause, which will cause a lot of confusion as it did to me while working with a bigger timeseries data.

If this bug is verified, from the reproducible script as attached, I'm happy to submit a PR for the same, thanks!
@phoeenniixx @fnhirwa @PranavBhatP

**Comment by FarahAbdo** [comment]
Hi @StrikerEureka34 , thanks for the detailed bug report and repro script!

I've submitted a fix in PR #2226 , it adds a recursive _move_to_cpu helper that calls .detach().cpu() on all tensors collected in on_predict_batch_end, covering:

- predictions (line ~69)
- return_info tensors (y, decoder_lengths)
- The x dict (handles mixed tensor/non-tensor values)

Happy to iterate based on reviewer feedback.

**Comment by fnhirwa** [comment]
hi @StrikerEureka34, nice catch, i was able to reproduce this and confirmed.

This proposed solution would work, and I believe you wanted to raise a PR for this, but given that @FarahAbdo has started working on the issue, perhaps you could provide the review on this PR, and we can work from that to get it into a mergeable state.

Thanks😊

**Comment by StrikerEureka34** [comment]
> This proposed solution would work, and I believe you wanted to raise a PR for this, but given that [@FarahAbdo](https://github.com/FarahAbdo) has started working on the issue, perhaps you could provide the review on this PR, and we can work from that to get it into a mergeable state.


Hi @fnhirwa , thanks for validating and jumping on this so quickly!
I have been working on a fix for this for a bit and have a version almost ready. 
I did find two things worth noting in @FarahAbdo 's RP:

1. No tests, there are no committed test cases, so I feel this could silently regress down the line.
2. There are existing utilities like pytorch_forecasting.utils that already exports detach and move_to_device, and the v1 base model uses exactly move_to_device(detach(out), "cpu") for the same purpose. A custom _move_to_cpu method ends up reinventing that.

Will share my PR shortly for comparison, thanks!

**Comment by StrikerEureka34** [comment]
Hi @FarahAbdo, I've just pushed my PR (#2228) and would love to compare notes! 
If you have any additions or improvements in mind feel free to bring them over, happy to combine efforts and work towards a mergeable state as @fnhirwa suggested. 
Looking forward to collaborating. thanks!
